### PR TITLE
Serializer_refactor - MOD-4196

### PR DESCRIPTION
**Pull request by meiravgri**
**Describe the changes in the pull request**

- In this PR we refactored the serializer class. it is **no longer holding an index instance** as a data member.

- If an index is supposed to be serialized it should **inherit the serializer** and **implement**:

1.  save to file function
2. A construct that initializes the index attributes from a file.

to avoid adding unnecessary functions and memory, the inheritance and serializing related functions can be declared and implemented under the **BUILD_TESTS** preprocessor flag. 

-When using encoding version 1 we need to explicitly generate the HNSWParams struct with the same attributes as the loaded index.
 - from encoding version 2 we serialize also `VecSimIndexAbstract` class data members (dim, vector data type, metric, etc.) so we can initialize HNSWParams for the ctor from the file.
 - 
**Which issues this PR fixes**
1. Resources management is better because we don't need to worry that the index will be freed before the serializer.
2. General API of the serializer enables serializing all types of indices in the future.
3. from v2 - we don't need to know the index attributes before loading it.

**Main objects this PR modified**
1. Serializer class is totally refactored and moved to the utils directory.
2. moved some serializing functions that are specific to hnsw index to hnsw class

**Mark if applicable**

- [ ] This PR introduces API changes
- [x] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 1]
```diff
@@ -331,7 +331,7 @@ VecSimIndexInfo BruteForceIndex<DataType, DistType>::info() const {
     info.bfInfo.indexSize = this->count;
     info.bfInfo.indexLabelCount = this->indexLabelCount();
     info.bfInfo.blockSize = this->blockSize;
-    info.bfInfo.memory = this->allocator->getAllocationSize();
+    info.bfInfo.memory = this->getAllocationSize();
```
Comment (meiravgri) [inline src/VecSim/algorithms/brute_force/brute_force.h:334]: getAllocationSize() was added to vecsim_interface

File: tests/benchmark/bm_utils.h [hunk 2]
```diff
@@ -32,13 +32,12 @@ class BM_VecSimBasics : public benchmark::Fixture {
...
-/*
- *  Populate the given hnsw_index with the serialized index data in the file
- *  which is located in the given path.
- */
-void load_HNSW_index(const char *path, VecSimIndex *hnsw_index);
+    static const char *hnsw_index_file;
+    static inline std::string GetSerializedIndexLocation(const char *file_name) {
+        return std::string(getenv("ROOT")) + "/" + file_name;
+    }
+};
```
Comment (meiravgri) [inline tests/benchmark/bm_utils.h:40]: serializer already includes stream opening logic

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 3]
```diff
@@ -1749,3 +1784,292 @@ bool HNSWIndex<DataType, DistType>::preferAdHocSearch(size_t subsetSize, size_t
...
+}
+
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::loadIndexIMP(std::ifstream &input, EncodingVersion version) {
+    // We already checked in the serializer that this is a valid version number.
+    if (version != EncodingVersion_V1) {
+        // This data is only serialized from V2 up.
+        VecSimAlgo algo = VecSimAlgo_BF;
+        readBinaryPOD(input, algo);
+        if (algo != VecSimAlgo_HNSWLIB) {
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw.h:1809]: cant do `VecSimAlgo algo;` because some architectures identify this as a condition on an initialized value and fail valgrind tests

File: src/VecSim/utils/serializer.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// Restore the index from the file in the specified location.
+void Serializer::loadIndex(const std::string &location) {
+    std::ifstream input(location, std::ios::binary);
+    if (!input.is_open()) {
+        throw std::runtime_error("Cannot open file");
+    }
+    input.seekg(0, std::ifstream::beg);
+
+    // The version number is the first field that is serialized.
+    EncodingVersion version = EncodingVersion_NOT_VALID;
```
Comment (meiravgri) [inline src/VecSim/utils/serializer.cpp:27]: cant do `EncodingVersion version;` because some architectures identify this as a condition on an initialized value and fail valgrind tests

File: tests/unit/test_hnsw_multi.cpp [hunk 5]
```diff
@@ -1733,3 +1632,98 @@ TYPED_TEST(HNSWMultiTest, rangeQuery) {
...
+    }
+
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n);
+    ASSERT_EQ(index->indexLabelCount(), n_labels);
+    // Persist index with the serializer, and delete it.
+    hnsw_index->saveIndex(this->file_name, Serializer::EncodingVersion_V2);
+
+    VecSimIndex_Free(index);
+
+    // Create new index, set it into the serializer and extract the data to it.
```
Comment (meiravgri) [inline tests/unit/test_hnsw_multi.cpp:1698]: ill fix the comment along with the review fixes

File: tests/unit/test_hnsw.cpp [hunk 6]
```diff
@@ -1838,6 +1840,198 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
+    hnsw_index->loadIndex(this->file_name);
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n);
+    ASSERT_TRUE(hnsw_index->serializingIsValid());
+
+    VecSimIndex_Free(index);
+}
+
 class HNSWGeneralTest : public ::testing::Test {};
 // TODO: serializer for double
 TEST_F(HNSWGeneralTest, hnsw_serialization_v1) {
```
Comment (meiravgri) [inline tests/unit/test_hnsw.cpp:2037]: this test is not templates because in v1 we don't save index type

File: tests/unit/test_hnsw.cpp [hunk 7]
```diff
@@ -1838,6 +1840,198 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
+                      .efRuntime = ef};
+    VecSimIndex *index = this->CreateNewIndex(params);
+    HNSWIndex<TEST_DATA_T, TEST_DIST_T> *hnsw_index = this->CastToHNSW(index);
+
+    this->GenerateFileName(std::string(getenv("ROOT")) +
+                           "/tests/unit/data/1k-d4-L2-M8-ef8_info.hnsw_v2");
+    // Save index.
+    EXPECT_THROW(hnsw_index->saveIndex(this->file_name, Serializer::EncodingVersion_NOT_VALID),
+                 std::runtime_error);
+    hnsw_index->saveIndex(this->file_name, Serializer::EncodingVersion_V2);
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1877]: You have this check already in test_common no?

File: tests/unit/test_hnsw.cpp [hunk 8]
```diff
@@ -1838,6 +1840,198 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
+    ASSERT_NE(serialized_index_info.hnswInfo.dim, other_index_info.hnswInfo.dim);
+    ASSERT_NE(serialized_index_info.hnswInfo.blockSize, other_index_info.hnswInfo.blockSize);
+    ASSERT_NE(serialized_index_info.hnswInfo.M, other_index_info.hnswInfo.M);
+    ASSERT_NE(serialized_index_info.hnswInfo.efConstruction,
+              other_index_info.hnswInfo.efConstruction);
+    ASSERT_NE(serialized_index_info.hnswInfo.efRuntime, other_index_info.hnswInfo.efRuntime);
+    ASSERT_NE(serialized_index_info.hnswInfo.epsilon, other_index_info.hnswInfo.epsilon);
+
+    hnsw_index = this->CastToHNSW(index);
+    hnsw_index->loadIndex(this->file_name);
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1902]: Looking at this test makes me think that we should use the `loadIndex` as an alternative constructor. Creating an index and then loading it from file was the temporary solution at the time, but now it looks awkward to use `loadIndex` to override index data. Let's discuss it...

**Review by alonre24 — CHANGES_REQUESTED** [review]
Good job! See the comments and let me know what you think. While most of the comments are minor, I do want us to discuss if we should change how we load index entirety, so that loading will be an alternative constructor for the index.

File: src/VecSim/utils/serializer.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+class Serializer {
+public:
+    typedef enum EncodingVersion {
+        EncodingVersion_V1 = 1,
+        EncodingVersion_V2 = 2,
+        EncodingVersion_NOT_VALID, // This should always be last.
+    } EncodingVersion;
+
+    // Persist index into a file in the specified location.
+    void saveIndex(const std::string &location, EncodingVersion version = EncodingVersion_V1);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/utils/serializer.h:16]: yes, i forgot to change this :smile:

File: src/VecSim/vec_sim_interface.h [hunk 10]
```diff
@@ -154,4 +155,6 @@ struct VecSimIndexInterface : public VecsimBaseObject {
     inline static void setTimeoutCallbackFunction(timeoutCallbackFunction callback) {
         VecSimIndexInterface::timeoutCallback = callback;
     }
+
+    inline int64_t getAllocationSize() const { return this->allocator->getAllocationSize(); }
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_interface.h:159]: this is just a shortcut

File: src/VecSim/algorithms/hnsw/hnsw_base_tests_friends.h [hunk 11]
```diff
@@ -6,3 +6,4 @@ INDEX_TEST_FRIEND_CLASS(HNSWMultiTest_preferAdHocOptimization_Test)
 INDEX_TEST_FRIEND_CLASS(HNSWMultiTest_MultiBatchIteratorHeapLogic_Test)
 INDEX_TEST_FRIEND_CLASS(IndexAllocatorTest_testIncomingEdgesSet_Test)
 INDEX_TEST_FRIEND_CLASS(IndexAllocatorTest_test_hnsw_reclaim_memory_Test)
+friend class SerializerTest_HNSWSerialzer_Test;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_base_tests_friends.h:9]: i guess not

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 12]
```diff
@@ -96,10 +107,34 @@ class HNSWIndex : public VecSimIndexAbstract<DistType> {
...
+    HNSWIndexMetaData checkIntegrity() const;
+    virtual void saveIndexIMP(std::ofstream &output, EncodingVersion version) const override;
+    virtual void loadIndexIMP(std::ifstream &input, EncodingVersion version) override;
+    virtual inline bool serializingIsValid() const override {
+        return this->checkIntegrity().valid_state;
+    }
+
+protected:
+    virtual void clearLabelLookup() = 0;
+    virtual void AddToLabelLookup(labelType label, idType id) = 0;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw.h:123]: no just implemented by them 
changed

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 13]
```diff
@@ -1749,3 +1784,292 @@ bool HNSWIndex<DataType, DistType>::preferAdHocSearch(size_t subsetSize, size_t
...
+    readBinaryPOD(input, this->max_elements_);
+    readBinaryPOD(input, this->M_);
+    readBinaryPOD(input, this->maxM_);
+    readBinaryPOD(input, this->maxM0_);
+    readBinaryPOD(input, this->ef_construction_);
+
+    // Restore index search parameter
+    readBinaryPOD(input, this->ef_);
+
+    // epsilon is only restored from v2 up.
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw.h:1916]: restoreIndexFields_v2

File: tests/unit/test_common.cpp [hunk 14]
```diff
@@ -359,3 +361,66 @@ TEST(CommonAPITest, VecSim_QueryResult_Iterator) {
...
+                      .initialCapacity = n,
+                      .blockSize = 1,
+                      .M = M,
+                      .efConstruction = ef,
+                      .efRuntime = ef};
+    VecSimIndex *index = test_utils::CreateNewIndex(params, VecSimType_FLOAT32);
+    HNSWIndex<float, float> *hnsw_index = reinterpret_cast<HNSWIndex<float, float> *>(index);
+
+    this->file_name = std::string(getenv("ROOT")) + "/tests/unit/data/test_common_hnsw";
+    ASSERT_EQ(this->GetFileSize(), 0);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_common.cpp:399]: i think i put this line here by mistake :sweat_smile:

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::HandleLevelGenerator(std::ifstream &input) {
+    if (this->m_version == EncodingVersion_V1) {
+        // All current v1 files were generated on intel machines, where
+        // sizeof(std::default_random_engine) ==  sizeof(unsigned long)
+        // unlike MacOS where sizeof(std::default_random_engine) ==  sizeof(unsigned int).
+
+        // Skip sizeof(unsigned long) bytes
+        input.ignore(sizeof(unsigned long));
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:166]: nice and clean `input.ignore()`!

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    // for V2 and up we don't serialize the level generator, so we just returb and
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:169]: typo `return`

File: tests/benchmark/bm_datasets.py [hunk 17]
```diff
@@ -29,7 +29,8 @@ def get_data_set(dataset_name):
 
 
 # Create an HNSW index from dataset based on specific params.
-def create_hnsw_index(dataset, ef_construction, M):
+def load_or_create_hnsw_index(dataset, index_file_name, ef_construction, M):
```
Comment (alonre24) [inline tests/benchmark/bm_datasets.py:32]: Did you ran this script locally to see that these changes work?

File: tests/unit/test_common.cpp [hunk 18]
```diff
@@ -359,3 +362,64 @@ TEST(CommonAPITest, VecSim_QueryResult_Iterator) {
...
+        file.seekg(0, std::ios::end);
+        const auto end = file.tellg();
+        file.close();
+
+        return end - begin;
+    }
+
+    std::string file_name;
+};
+TEST_F(SerializerTest, HNSWSerialzer) {
```
Comment (alonre24) [inline tests/unit/test_common.cpp:382]:
```suggestion
TEST_F(SerializerTest, HNSWSerialzerErrors) {
```
Also, try to see if you can increase the coverage in some additional cases where we send invalid `params` to the new constructor (bad `M`, bad type etc...), and test the error message itself in addition to the exception type

File: tests/unit/test_hnsw.cpp [hunk 19]
```diff
@@ -1838,103 +1840,166 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
 
-    HNSWIndex<float, float> *hnswNewIndex = reinterpret_cast<HNSWIndex<float, float> *>(new_index);
-    ASSERT_EQ(hnswNewIndex->getIndexCapacity(), n);
+        VecSimIndex_Free(index);
 
-    // Persist index, delete it from memory and restore.
-    serializer.saveIndex(file_name);
-    VecSimIndex_Free(new_index);
+        // Load the index from the file.
+        VecSimIndex *serialized_index = HNSWFactory::NewIndex(file_name, &params);
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1907]: Why sending `params`? I thought that for V2 you don't need them.

File: tests/unit/test_hnsw.cpp [hunk 20]
```diff
@@ -1838,103 +1840,166 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
-    res = serializer.checkIntegrity();
-    ASSERT_TRUE(res.valid_state);
+    size_t dim = 4;
+    size_t n = 1000;
+    size_t n_labels[] = {n, 100};
+    size_t M = 8;
+    size_t ef = 10;
+    double epsilon = 0.004;
+    size_t blockSize = 1;
+    bool is_multi[] = {false, true};
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:1953]: We have `multi` in v1?

**Review by alonre24 — COMMENTED** [review]
Great job! :)
See some comments and questions that I had. Also, I would add some more tests such as - serializing and loading an empty index, searching before and after serialization (only in v2...)

File: src/VecSim/utils/serializer.cpp [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+#include <string>
+
+#include "VecSim/utils/serializer.h"
+
+// Persist index into a file in the specified location.
+void Serializer::saveIndex(const std::string &location, EncodingVersion version) {
+    // Only V1 and V2 are supported
+    if (version >= EncodingVersion_INVALID) {
+        throw std::runtime_error("Cannot load index: bad encoding version");
+    }
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/utils/serializer.cpp:11]: you mean to fix the comment?

File: src/VecSim/vec_sim_index.h [hunk 22]
```diff
@@ -42,6 +42,12 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
         spaces::SetDistFunc(metric, dim, &dist_func);
     }
 
+#ifdef BUILD_TESTS
+    VecSimIndexAbstract(std::shared_ptr<VecSimAllocator> allocator)
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_index.h:46]: i didn't use it eventually

File: src/VecSim/algorithms/hnsw/hnsw_serializer_declarations.h [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+// Serializing and tests functions.
+public:
+HNSWIndex(std::ifstream &input, const HNSWParams *params,
+          std::shared_ptr<VecSimAllocator> allocator, EncodingVersion version);
+
+// Validates the connections between vectors
+HNSWIndexMetaData checkIntegrity() const;
+
+// Index memory size might be changed during index saving.
+virtual void saveIndexIMP(std::ofstream &output, EncodingVersion version) override;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_serializer_declarations.h:12]: it is pure virtual in the serializer class so it can call it from saveIndex function

File: tests/benchmark/bm_datasets.py (hunk 17 above)
Comment (meiravgri) [reply tests/benchmark/bm_datasets.py:32]: yes i ran make benchmark

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 24]
```diff
@@ -119,4 +119,84 @@ size_t EstimateElementSize(const HNSWParams *params) {
...
+    }
+    case Serializer::EncodingVersion_V1: {
+        assert(params);
+        file_params.type = params->type;
+        file_params.dim = params->dim;
+        file_params.metric = params->metric;
+        file_params.multi = params->multi;
+        file_params.blockSize = params->blockSize;
+        file_params.epsilon = params->epsilon;
+        break;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:182]: Good question
maybe epsilon and block size can have default values because they don't affect the loading process, but all the others must be the same as the loaded index since we cant get them from the file.

File: tests/unit/test_hnsw.cpp (hunk 20 above)
Comment (meiravgri) [reply tests/unit/test_hnsw.cpp:1953]: Theoretically we can have

File: tests/unit/test_hnsw.cpp (hunk 19 above)
Comment (meiravgri) [reply tests/unit/test_hnsw.cpp:1907]: remvoed

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+            for (size_t l = 0; l <= this->element_levels_[i]; l++) {
+                linklistsizeint *ll_cur = this->get_linklist_at_level(i, l);
+                unsigned int size = this->getListCount(ll_cur);
+                auto *data = (idType *)(ll_cur + 1);
+                std::set<idType> s;
+                for (unsigned int j = 0; j < size; j++) {
+                    // Check if we found an invalid neighbor.
+                    if (data[j] > this->max_id || data[j] == i) {
+                        res.valid_state = false;
+                        return res;
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:79]: was never covered

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+                        if (data_other[r] == (idType)i) {
+                            double_connections++;
+                            break;
+                        }
+                    }
+                }
+                // Check if a certain neighbor appeared more than once.
+                if (s.size() != size) {
+                    res.valid_state = false;
+                    return res;
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:99]: was never covered

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+        !inbound_connections_num.empty()
+            ? *std::min_element(inbound_connections_num.begin(), inbound_connections_num.end())
+            : 0;
+    res.max_in_degree =
+        !inbound_connections_num.empty()
+            ? *std::max_element(inbound_connections_num.begin(), inbound_connections_num.end())
+            : 0;
+    if (incoming_edges_sets_sizes + double_connections != connections_checked) {
+        res.valid_state = false;
+        return res;
```
Comment (meiravgri) [inline src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:117]: was never covered

**Review by alonre24 — APPROVED** [review]
You did it! :)

**Review by GuyAv46 — COMMENTED** [review]
Nicely done!
regarding adding invalid values to enums, IMO we should avoid it

File: src/VecSim/vec_sim_interface.h [hunk 28]
```diff
@@ -1,10 +1,11 @@
 #pragma once
 #include "vec_sim_common.h"
 #include "query_results.h"
-#include <stddef.h>
 #include "VecSim/memory/vecsim_base.h"
 #include "info_iterator_struct.h"
 
+#include <stddef.h>
+#include <stdexcept>
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_interface.h:8]: don't remember but I removed it and it's ok

File: src/VecSim/algorithms/hnsw/hnsw_multi.h [hunk 29]
```diff
@@ -10,8 +10,14 @@ class HNSWIndex_Multi : public HNSWIndex<DataType, DistType> {
...
 
 #ifdef BUILD_TESTS
-    friend class HNSWIndexSerializer;
 #include "VecSim/algorithms/hnsw/hnsw_multi_tests_friends.h"
+    virtual void AddToLabelLookup(labelType label, idType id) override {
+        if (label_lookup_.find(label) == label_lookup_.end()) {
+            label_lookup_.emplace(label, vecsim_stl::vector<idType>{this->allocator});
+        }
+        setVectorId(label, id);
+    }
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_multi.h:19]: this function is called by the parent class so if we dont want to affect performance when we are not in tests mode we need the virtual function wrapper anyway

File: tests/unit/test_hnsw.cpp [hunk 30]
```diff
@@ -1838,103 +1840,182 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
...
-    std::mt19937 rng;
-    rng.seed(47);
-    std::uniform_real_distribution<> distrib;
-    for (size_t i = 0; i < (n + 1) * dim; ++i) {
-        data[i] = (float)distrib(rng);
-    }
-    for (size_t i = 0; i < n + 1; ++i) {
-        VecSimIndex_AddVector(new_index, data.data() + dim * i, i);
-    }
+        auto file_name = std::string(getenv("ROOT")) + "/tests/unit/data/1k-d4-L2-M8-ef_c10_" +
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnsw.cpp:1885]: couldn't find any function that is platform-independent to get pwd... (do you know any?)
and "."doesn't work because we are running tests from the dir
we define ROOT in the cmake file, right?
but yes anyway it won't work using the debugger unless you explicitly config ROOT

File: src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h [hunk 31]
```diff
@@ -0,0 +1,307 @@
...
+    writeBinaryPOD(output, this->metric);
+    writeBinaryPOD(output, this->blockSize);
+    writeBinaryPOD(output, this->isMulti);
+    writeBinaryPOD(output, this->epsilon_);
+}
+
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::saveIndexFields(std::ofstream &output) const {
+
+    this->saveIndexFields_v2(output);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_serializer_imp.h:239]: yeah we always call both but I think it's not a bad idea to put them under different "headlines"
From a readability point of view

File: src/VecSim/utils/serializer.cpp [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+
+Serializer::EncodingVersion Serializer::ReadVersion(std::ifstream &input) {
+
+    input.seekg(0, std::ifstream::beg);
+
+    // The version number is the first field that is serialized.
+    EncodingVersion version = EncodingVersion_INVALID;
+    readBinaryPOD(input, version);
+    // Only V1 and V2 are supported
+    if (version >= EncodingVersion_INVALID) {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/utils/serializer.cpp:26]: you're right @meiravgri. But we don't return an error if `version==0`. You can just add this validation to this `if`

File: src/VecSim/algorithms/hnsw/hnsw_multi.h (hunk 29 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_multi.h:19]: It won’t effect performance because it is called once every insertion anyway

File: tests/unit/test_hnsw.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw.cpp:1885]: `std::filesystem::current_path()` should work. Maybe use relative path to `data/`

File: src/VecSim/utils/serializer.cpp (hunk 32 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/utils/serializer.cpp:26]: EncodingVersion==0 won't throw because it is smaller than EncodingVersion_INVALID
`    typedef enum EncodingVersion {
        EncodingVersion_V1 = 1,
        EncodingVersion_V2 = 2,
        EncodingVersion_INVALID, // This should always be last.
    } EncodingVersion;`
I can also add EncodingVersion_V0,
but it will return when we try to initialize params. So I don't quite understand how to handle v0 without returning an error

[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/utils/serializer.cpp:26]: I think you SHOULD return an error for `version==0`

[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/serializer.cpp:26]: Agrees. 
```suggestion
    // Only V1 and V2 are supported
    if (version <= 0 || version >= EncodingVersion_INVALID) {
```

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/utils/serializer.cpp:26]: oh got it :sweat_smile:

File: src/VecSim/vec_sim_common.h [hunk 33]
```diff
@@ -18,11 +18,12 @@ typedef enum {
...
     VecSimType_FLOAT64,
     VecSimType_INT32,
-    VecSimType_INT64
+    VecSimType_INT64,
+    VecSimType_INVALID
 } VecSimType;
 
 // Algorithm type/library.
-typedef enum { VecSimAlgo_BF, VecSimAlgo_HNSWLIB } VecSimAlgo;
+typedef enum { VecSimAlgo_BF, VecSimAlgo_HNSWLIB, VecSimAlgo_INVALID } VecSimAlgo;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_common.h:26]: we need this to intialize variables before loading the file data into it
fo rexample:
`        VecSimAlgo algo = VecSimAlgo_INVALID;
        Serializer::readBinaryPOD(input, algo)`

in hnsw_factory

File: src/VecSim/vec_sim.cpp [hunk 34]
```diff
@@ -119,8 +119,9 @@ extern "C" size_t VecSimIndex_EstimateInitialSize(const VecSimParams *params) {
         return HNSWFactory::EstimateInitialSize(&params->hnswParams);
     case VecSimAlgo_BF:
         return BruteForceFactory::EstimateInitialSize(&params->bfParams);
+    default:
+        return -1;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim.cpp:123]: what's the difference between this and just returning -1 if its not HNSW or BF?

File: tests/benchmark/bm_basics.cpp [hunk 35]
```diff
@@ -69,12 +68,11 @@ BENCHMARK_DEFINE_F(BM_VecSimBasics, DeleteVectorHNSW)(benchmark::State &st) {
...
     double memory_delta = 0;
     size_t iter = 0;
+    auto hnsw_index_casted = reinterpret_cast<HNSWIndex<float, float> *>(hnsw_index);
     for (auto _ : st) {
         st.PauseTiming();
         auto removed_vec = std::vector<float>(dim);
-        memcpy(removed_vec.data(),
-               reinterpret_cast<HNSWIndex<float, float> *>(hnsw_index)
-                   ->getDataByInternalId(id_to_remove),
+        memcpy(removed_vec.data(), hnsw_index_casted->getDataByInternalId(id_to_remove),
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/benchmark/bm_basics.cpp:75]: not really i guess

File: tests/benchmark/bm_updated_index.cpp [hunk 36]
```diff
@@ -47,33 +48,36 @@ class BM_VecSimUpdatedIndex : public BM_VecSimBasics {
...
-                                   HNSWParams{.type = VecSimType_FLOAT32,
-                                              .dim = BM_VecSimBasics::dim,
-                                              .metric = VecSimMetric_Cosine,
-                                              .initialCapacity = BM_VecSimBasics::n_vectors,
-                                              .M = BM_VecSimBasics::M,
-                                              .efConstruction = BM_VecSimBasics::EF_C}};
-        BM_VecSimUpdatedIndex::hnsw_index_updated = VecSimIndex_New(&params);
-        load_HNSW_index(updated_hnsw_index_file, hnsw_index_updated);
+        // Generate the updated index from file.
+        // HNSWParams is required to load v1 index
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/benchmark/bm_updated_index.cpp:60]: passing the params struct part will be removed

File: src/VecSim/vec_sim_interface.h (hunk 28 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_interface.h]: ok i don't why it doesn't crush locally but the default flow it says that the it doesn't know std::runtime_error

**Review by GuyAv46 — APPROVED** [review]
